zerver/lib/retention.py

```diff
@@ -502,6 +522,172 @@ def move_messages_to_archive(
         raise Message.DoesNotExist
 
 
+def extract_dm_group_archive_data_by_message_ids(
+    message_ids: list[int],
+    chunk_size: int = MESSAGE_BATCH_SIZE,
+) -> dict[int, DirectMessageGroupArchiveData]:
+    archive_data: dict[int, DirectMessageGroupArchiveData] = {}
+
+    # Process in chunks to avoid overwhelming the database
+    message_ids_head = message_ids
+    while message_ids_head:
+        message_ids_chunk = message_ids_head[0:chunk_size]
+        message_ids_head = message_ids_head[chunk_size:]
+
+        affected_recipients = (
+            Message.objects.select_related("recipient")
+            .filter(
+                id__in=message_ids_chunk,
+                is_channel_message=False,
+                recipient__type=Recipient.DIRECT_MESSAGE_GROUP,
+            )
+            .values_list("recipient_id", "recipient__type_id")
+        )
+
+        for recipient_id, dm_group_id in affected_recipients:
+            if dm_group_id not in archive_data:
+                archive_data[dm_group_id] = DirectMessageGroupArchiveData(
+                    recipient_id=recipient_id,
+                    archived_message_count=0,
+                )
+            archive_data[dm_group_id].archived_message_count += 1
+
+    return archive_data
+
+
+def extract_dm_group_archive_data_by_check_date(
+    realm: Realm,
+    check_date: datetime,
+    recipient: Recipient | None = None,
+) -> tuple[dict[int, DirectMessageGroupArchiveData], list[int]]:
+    archive_data: dict[int, DirectMessageGroupArchiveData] = {}
+    archived_message_ids: list[int] = []
+
+    query = Message.objects.filter(realm_id=realm.id, date_sent__lt=check_date)
+    if recipient:
+        assert recipient.type == Recipient.DIRECT_MESSAGE_GROUP
+        query = query.filter(recipient_id=recipient.id)
+    else:
+        query = query.prefetch_related("recipient").filter(
+            is_channel_message=False,
+            recipient__type=Recipient.DIRECT_MESSAGE_GROUP,
+        )
```

Comment:
[The same comment](https://github.com/zulip/zulip/pull/36045/files#r2347033716).